File: 백준/Gold/16973. 직사각형 탈출/직사각형 탈출.py
import sys
from collections import deque

input = sys.stdin.readline

n, m = map(int, input().split())

graph = []
wall = []
for i in range(n):
    e = list(map(int, input().split()))
    for j in range(m):
        if e[j] == 1:
            wall.append((i, j))

h, w, sr, sc, fr, fc = map(int, input().split())
sr -= 1
sc -= 1
fr -= 1
fc -= 1


def bfs():
    visit = [[0] * (m) for _ in range(n)]
    visit[sr][sc] = 1
    q = deque()
    q.append((sr, sc, 0))
    dx = [-1, 0, 1, 0]
    dy = [0, 1, 0, -1]
    while q:
        a1, a2, a3 = q.popleft()
        if a1 == fr and a2 == fc:
            return a3
        for i in range(4):
            zx = a1 + dx[i]
            zy = a2 + dy[i]
            if 0 <= zx < n and 0 <= zy < m and visit[zx][zy] == 0 and zx + h <= n and zy + w <= m:
                # 벽이 있는지 없는지 판단 해아함

                cc = 0
                for z1, z2 in wall:
                    if zx <= z1 < zx + h and zy <= z2 < zy + w:
                        cc = 1
                        break
                if cc == 0:
                    visit[zx][zy] = 1
                    q.append((zx, zy, a3 + 1))
    return -1

zz = bfs()
if zz != -1:
    print(zz)
else:
    print(-1)

File: 백준/Gold/16973. 직사각형 탈출/직사각형 탈출.java


import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayDeque;
import java.util.Deque;
import java.util.StringTokenizer;

public class Main {


    static int n, m;
    static int[][] graph;
    static int[][] add;
    static int h, w, sr, sc, fr, fc;


    static class go {
        int x, y;

        go(int x, int y) {
            this.x = x;
            this.y = y;
        }
    }

    static int check(int zx, int zy) {
        if (1 <= zx && zx <= n && 1 <= zy && zy <= m)
            return 1;
        return -1;
    }

    public static int bfs() {

        Deque<go> q = new ArrayDeque<>();
        int[][] visit = new int[n + 2][m + 2];
        visit[sr][sc] = 1;
        int dx[] = {-1, 0, 1, 0};
        int dy[] = {0, -1, 0, 1};
        q.add(new go(sr, sc));

        while (!q.isEmpty()) {
            go a = q.poll();
            if (a.x == fr && a.y == fc) {
                return visit[fr][fc] - 1;
            }
            for (int i = 0; i < 4; i++) {
                int zx = dx[i] + a.x;
                int zy = dy[i] + a.y;
                int tx = zx + h - 1;
                int ty = zy + w - 1;
                if (check(zx, zy) == 1) {
                    if (visit[zx][zy] !=0 || check(tx, ty) != 1)
                        continue;
                    //이제 벽이 있는지 확인해 봐야함
                    int temp = 0;
                    for (int i1 = zx; i1 < zx + h; i1++) {
                        temp += add[i1][ty] - add[i1][zy - 1];
                        if (temp > 0)
                            break;
                    }
                    if (temp == 0) {
                        visit[zx][zy] += visit[a.x][a.y] + 1;
                        q.add(new go(zx, zy));
                    }
                }
            }
        }


        return -1;

    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine(), " ");
        n = Integer.parseInt(st.nextToken());
        m = Integer.parseInt(st.nextToken());
        graph = new int[n + 2][m + 2];
        add = new int[n + 2][m + 2];
        for (int i = 1; i <= n; i++) {
            st = new StringTokenizer(br.readLine(), " ");
            for (int j = 1; j <= m; j++) {
                graph[i][j] = Integer.parseInt(st.nextToken());
                add[i][j] = graph[i][j];
            }
            for (int j = 2; j <= m; j++) {
                add[i][j] += add[i][j - 1];
            }
        }
        st = new StringTokenizer(br.readLine(), " ");
        h = Integer.parseInt(st.nextToken());
        w = Integer.parseInt(st.nextToken());
        sr = Integer.parseInt(st.nextToken());
        sc = Integer.parseInt(st.nextToken());
        fr = Integer.parseInt(st.nextToken());
        fc = Integer.parseInt(st.nextToken());

        System.out.println(bfs());

    }


}
